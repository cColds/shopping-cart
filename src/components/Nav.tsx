import logo from "../assets/logo.svg";
import { Link } from "react-router-dom";
import { FaSearch, FaShoppingCart } from "react-icons/fa";
import { useState, ChangeEvent } from "react";
import Item from "../interfaces/Item";

export default function Nav({
  itemCount,
  onCartToggle,
  items,
}: {
  itemCount: number;
  onCartToggle: () => void;
  items: Item[];
}) {
  const [isSearchOpen, setIsSearchOpen] = useState(false);
  const [searchResults, setSearchResults] = useState<Item[]>([]);
  const [searchValue, setSearchValue] = useState("");

  const handleSearchToggle = () => {
    if (!isSearchOpen) {
      setSearchResults([]);
      setSearchValue("");
    }
    setIsSearchOpen(!isSearchOpen);
  };
  const handleSearchInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    const updatedSearchValue = e.target.value.toLowerCase();
    setSearchValue(updatedSearchValue);
    if (updatedSearchValue.trim() === "") {
      setSearchResults([]);
      return;
    }

    const updatedSearchResults = items.filter((item) => {
      const { title } = item.node;
      if (title.toLowerCase().includes(updatedSearchValue.trim())) return item;
    });

    setSearchResults(updatedSearchResults);
  };

  return (
    <nav className="flex items-center justify-between gap-3 px-4 py-2	shadow-md">
      <Link to="/" aria-label="home page" className="shrink-0">
        <img src={logo} alt="fashonique logo" className="h-[30px] w-[150px]" />
      </Link>
      <button
        type="button"
        className="flex h-[40px] w-[250px] items-center gap-2 rounded-lg border-[1px] border-slate-200 focus:ring-sky-200"
        onClick={handleSearchToggle}
      >
        <FaSearch className="m-2 mr-0 text-slate-500" aria-hidden="true" />
        <span className="m-2 ml-0 text-sm text-slate-500">Search</span>
      </button>

      <div
        data-search-modal={isSearchOpen}
        className="fixed right-0 top-0 z-50 flex h-full w-full items-center justify-center bg-modal backdrop-blur-sm data-[search-modal='false']:invisible"
        onClick={handleSearchToggle}
        onKeyDown={(e) => {
          if (e.key === "Escape") handleSearchToggle();
        }}
      >
        <div
          className="flex h-[450px] w-[450px] flex-col rounded-lg bg-white"
          onClick={(e) => e.stopPropagation()}
        >
          <form
            role="search"
            className="flex h-[60px] w-full items-center border-b-[1px] border-slate-200 p-1"
          >
            <FaSearch className="m-2 text-slate-500" aria-hidden="true" />
            <input
              type="search"
              id="search"
              placeholder="Search"
              autoComplete="off"
              className="h-full w-full focus:ring-0"
              onChange={handleSearchInputChange}
              value={searchValue}
            />
          </form>

          <ul className="flex flex-col gap-2 overflow-auto p-2.5">
            <h1 className="text-lg">
              {searchResults.length
                ? `Items (${searchResults.length})`
                : `No results found`}
            </h1>
            {searchResults.map((item) => {
              const { title } = item.node;
              const itemUrl = title.replace(/\s+/g, "-").toLowerCase();
              return (
                <li key={item.node.id} className="flex">
                  <a
                    className="flex h-[50px] flex-grow items-center rounded-lg border-b-[1px]  border-slate-200 bg-slate-100 p-2.5 text-sm"
                    href={`/store/${itemUrl}`}
                  >
                    {title}
                  </a>
                </li>
              );
            })}
          </ul>
        </div>
      </div>

      <ul className="flex gap-5">
        <li className="flex items-center">
          <Link to="/">Home</Link>
        </li>
        <li className="flex items-center">
          <Link to="/store">Store</Link>
        </li>
        <li className="flex">
          <button className="relative" onClick={onCartToggle}>
            <FaShoppingCart className="text-xl" />
            <span className="absolute right-[-10px] top-[-5px] flex h-[16px] min-w-[16px] items-center justify-center rounded-full bg-red-600 p-0.5 text-xs text-white">
              {itemCount < 100 ? itemCount : 99 + "+"}
            </span>
          </button>
        </li>
      </ul>
    </nav>
  );
}
