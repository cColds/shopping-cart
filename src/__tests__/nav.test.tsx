import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import Nav from "../components/Nav";

it("logo should link to home page", () => {
  render(
    <MemoryRouter>
      <Nav />
    </MemoryRouter>
  );
  const logoLink = screen.getByRole("link", { name: /home page/i });
  expect(logoLink).toHaveAttribute("href", "/");
});

it("nav links path should be correct", () => {
  render(
    <MemoryRouter>
      <Nav />
    </MemoryRouter>
  );
  const homeLink = screen.getByRole("link", { name: /^home$/i });
  const storeLink = screen.getByRole("link", { name: /^store$/i });
  expect(homeLink).toHaveAttribute("href", "/");
  expect(storeLink).toHaveAttribute("href", "/store");
});

it("should render nav and match snapshot", () => {
  const { container } = render(
    <MemoryRouter>
      <Nav />
    </MemoryRouter>
  );
  expect(container).toMatchSnapshot();
});
