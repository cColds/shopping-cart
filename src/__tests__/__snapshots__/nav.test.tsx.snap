// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders nav 1`] = `
<div>
  <div
    class="flex justify-between px-6 py-2 gap-3 items-center	shadow-md"
  >
    <a
      aria-label="home page"
      class="shrink-0"
      href="/"
    >
      <img
        alt="fashonique logo"
        class="w-[200px] h-[30px]"
        src="/src/assets/logo.svg"
      />
    </a>
    <form
      class="flex items-center border-[1px] border-slate-200 rounded-lg h-[40px] w-[min-content]"
      role="search"
    >
      <button
        class="m-2"
      >
        <svg
          class="text-slate-500"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          />
        </svg>
      </button>
      <input
        autocomplete="off"
        class="p-1 bg-transparent"
        id="search"
        placeholder="Search"
        type="search"
      />
    </form>
    <ul
      class="flex gap-3 text-lg"
    >
      <li>
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li>
        <a
          href="/store"
        >
          Store
        </a>
      </li>
      <li>
        <a
          href="/#contact"
        >
          Contact
        </a>
      </li>
      <li
        class="flex"
      >
        <button>
          <svg
            class="text-xl"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M528.12 301.319l47.273-208C578.806 78.301 567.391 64 551.99 64H159.208l-9.166-44.81C147.758 8.021 137.93 0 126.529 0H24C10.745 0 0 10.745 0 24v16c0 13.255 10.745 24 24 24h69.883l70.248 343.435C147.325 417.1 136 435.222 136 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-15.674-6.447-29.835-16.824-40h209.647C430.447 426.165 424 440.326 424 456c0 30.928 25.072 56 56 56s56-25.072 56-56c0-22.172-12.888-41.332-31.579-50.405l5.517-24.276c3.413-15.018-8.002-29.319-23.403-29.319H218.117l-6.545-32h293.145c11.206 0 20.92-7.754 23.403-18.681z"
            />
          </svg>
        </button>
      </li>
    </ul>
  </div>
</div>
`;
